family/src/mcl/mcl/doc/mcl.1
.\" Copyright (c) 2002 Stijn van Dongen
.TH mcl 1 "22 Feb 2002" "mcl 1\&.00" "USER COMMANDS"
.SH NAME

mcl \- the Amsterdam implementation of the Markov Cluster Algorithm
.SH SYNOPSIS

\fBmcl\fP <-|fname>
\fB[-I\fP f (\fIinflation\fP)\fB]\fP
\fB[-o\fP str (\fIfname\fP)\fB]\fP
.br
These two options are sufficient in 95 percent of the cases, or
more\&.

\fBmcl\fP <-|fname>
\fB[-I\fP f (\fIinflation\fP)\fB]\fP
\fB[-o\fP str (\fIfname\fP)\fB]\fP
\fB[-c\fP f (\fIcentering\fP)\fB]\fP
\fB[-p\fP f (\fIcutoff\fP)\fB]\fP
\fB[-P\fP n (\fI1/cutoff\fP)\fB]\fP
\fB[-S\fP n (\fIselection number\fP)\fB]\fP
\fB[-R\fP n (\fIrecovery number\fP)\fB]\fP
\fB[-pct\fP f (\fIrecover percentage\fP)\fB]\fP
\fB[-scheme\fP k (\fIuse preset scheme\fP)\fB]\fP
\fB[--show-schemes\fP (\fIshow preset schemes\fP)\fB]\fP
\fB[-warn-pct\fP n (\fIprune warn percentage\fP)\fB]\fP
\fB[-warn-factor\fP n (\fIprune warn factor\fP)\fB]\fP
\fB[--rigid\fP (\fIpruning\fP)\fB]\fP
\fB[-ae\fP f (\fIadaptive pruning exponent\fP)\fB]\fP
\fB[-af\fP f (\fIadaptive pruning factor\fP)\fB]\fP
\fB[--adapt\fP (\fIpruning\fP)\fB]\fP
\fB[-nx\fP x (\fItrack worst n\fP)\fB]\fP
\fB[-ny\fP y (\fItrack worst n\fP)\fB]\fP
\fB[-v\fP str (\fIverbosity type on\fP)\fB]\fP
\fB[-V\fP str (\fIverbosity type off\fP)\fB]\fP
\fB[--silent\fP (\fIvery\fP)\fB]\fP
\fB[--verbose\fP (\fIvery\fP)\fB]\fP
\fB[-progress\fP k (\fIgauge\fP)\fB]\fP
\fB[-te\fP k (\fI#expansion threads\fP)\fB]\fP
\fB[-ti\fP k (\fI#inflation threads\fP)\fB]\fP
\fB[--clone\fP (\fIwhen threading\fP)\fB]\fP
\fB[-cloneat\fP n (\fItrigger\fP)\fB]\fP
\fB[-t\fP k (\fI#threads\fP)\fB]\fP
\fB[-l\fP n (\fIinitial iteration number\fP)\fB]\fP
\fB[-L\fP n (\fImain iteration number\fP)\fB]\fP
\fB[-i\fP f (\fIinitial inflation\fP)\fB]\fP
\fB[-a\fP f (\fIloop weight\fP)\fB]\fP
\fB[-dumpi\fP i:j (\fIinterval i\&.\&.j-1\fP)\fB]\fP
\fB[-dumpm\fP k (\fIdump i+0\&.\&.i+k\&.\&.\fP)\fB]\fP
\fB[-dumpstem\fP stem (\fIfile stem\fP)\fB]\fP
\fB[-dump\fP str (\fItype\fP)\fB]\fP
\fB[-digits\fP n (\fIprinting precision\fP)\fB]\fP
\fB[--show\fP (\fIprint matrices to screen\fP)\fB]\fP
\fB[--ascii\fP (\fIoutput format\fP)\fB]\fP
\fB[--binary\fP (\fIoutput format\fP)\fB]\fP
\fB[--overlap\fP (\fIkeep it\fP)\fB]\fP
\fB[--expand-only\fP (\fIfactor out computation\fP)\fB]\fP
\fB[--inflate-first\fP (\fIrather then expand\fP)\fB]\fP
\fB[-preprune\fP n (\fIinput matrix\fP)\fB]\fP
\fB[-z\fP (\fIshow current settings\fP)\fB]\fP
.SH DESCRIPTION

\fBmcl\fP implements the \fBMCL algorithm\fP, short for the \fBMarkov cluster
algorithm\fP, a cluster algorithm for graphs developed by Stijn van Dongen at
the Centre for Mathematics and Computer Science in Amsterdam, the
Netherlands\&. The algorithm simulates flow using two simple algebraic
operations on matrices\&. The theory behind it is extensively described
elsewhere (see \fBREFERENCES\fP)\&. The program described here is a fast
threaded implementation written by the algorithm\&'s creator with
contributions by several others\&. Anton Enright co-implemented threading; see
the \fBHISTORY/CREDITS\fP section for a complete account\&. The implementation is
used for the TRIBES project in which large numbers of proteins are clustered
into families, and has become all the better from the feedback this has
generated\&.
See the \fBAPPLICABILITY\fP section for a description of the type of
graph \fBmcl\fP likes best, and for a qualitative assessment of its speed\&.

The \fB-I\fP\ \fIf\fP option is the main control,
affecting cluster granularity\&. Using \fBmcl\fP is as simple as
typing (assuming a file \fIproteins\fP contains a matrix/graph
in \fBmcl\fP input format)

.nf \fC
   mcl proteins -I 2\&.0
.fi \fR

The above will result in a clustering written to the file
named \fIout\&.mcl\fP\&. The \fBmcl\fP input format is described in the
\fBmcx(5)\fP section\&. Clusterings are also stored as matrices
- this is again discussed in the \fBmcx(5)\fP section\&.
In finding good \fBmcl\fP parameter settings for a particular domain,
or in finding cluster structure at different levels of granularity,
one typically runs \fBmcl\fP multiple times for varying values of f (refer
to the \fB-I\fP option for further information)\&.

\fBmcl\fP expects a nonnegative matrix in the input file, or equivalently, a
weighted (possibly directed) graph\&. NOTE \- \fBmcl\fP interprets the matrix
entries or graph edge weights as \fBsimilarities\fP, and it likes
\fBundirected input graphs\fP best\&. It can handle directed graphs, but any
node pair (i,j) for which w(i,j) is much smaller than w(j,i) or vice versa
will presumably have a slightly negative effect on the clusterings output by
\fBmcl\fP\&. Many such node pairs will have a distinctly negative effect, so try to
make your input graphs undirected\&. How your edge weights are computed may
affect \fBmcl\fP\&'s performance\&. In protein clustering, it is best to choose the
negated logarithm of the BLAST probabilities (see \fBREFERENCES\fP)\&.

\fBmcl\fP\&'s default parameters should make it quite fast in almost all
circumstances\&. Taking default parameters, \fBmcl\fP has been used to generate
good protein clusters on 133k proteins, taking 10 minutes running time on a
Compaq ES40 system with four alpha EV6\&.7 processors\&.

For large graphs, there are several groups of parameters available for
tuning the mcl computing process, should it be necessary\&.
The easiest thing
to do is just vary the \fB-scheme\fP option\&. This
triggers different settings for the group of pruning parameters
\fB{\fP\fB-p/-P\fP, \fB-R\fP, \fB-S\fP, and
\fB-pct\fP\fB}\fP\&. The default setting corresponds with
\fB-scheme\fP\ \fB2\fP\&.
There is an additional group of control parameters
\fB{\fP\fB--adapt\fP, \fB--rigid\fP, \fB-ae\fP,
\fB-af\fP\fB}\fP, which may be helpful in speeding up \fBmcl\fP\&.
When doing multiple \fBmcl\fP runs for the same graphs with different
\fB-I\fP settings (for obtaining clusterings at different levels
of granularity), it can be useful to factor out the first bit
of computation that is common to all runs, by using
the \fB--expand-only\fP option one time
and then using \fB--inflate-first\fP for
each run in the set\&.
Whether \fBmcl\fP considers a graph large depends mainly on the graph
connectivity; a highly connected graph on 50,000 nodes is large to
\fBmcl\fP (so that you might want to tune resources) whereas a sparsely
connected graph on 500,000 nodes may be business as usual\&. If graphs
are really huge, the time to read a graph from file can be shortened
by converting the input graph from ascii mcl format to binary mcl
format with \fBmcxconvert(1)\fP\&.

Two other groups of interest are the thread-related
options (you can specify the number of threads to use)
\fB{\fP\fB-t\fP, \fB-te\fP, \fB-ti\fP,
\fB--clone\fP, \fB-cloneat\fP\fB}\fP
and the verbosity-related options
\fB{\fP\fB--verbose\fP, \fB--silent\fP, \fB-v\fP,
\fB-V\fP\fB}\fP\&.
The actual settings are shown with \fB-z\fP, and for graphs with
at most 12 nodes or so you can view the MCL matrix iterands on screen
by supplying \fB--show\fP (this may give some
more feeling)\&.

The first option is the input file name (see the \fBmcx(5)\fP section
for its expected format), or a single hyphen to read from stdin\&.
The rationale is that you typically do several runs with different
parameters, and in command line mode it is pleasant if you do not have
to skip over an immutable parameter all the time\&.

In the \fBOPTIONS\fP section options are listed in order of
importance, with related options grouped together\&.

The creator of this page feels that manual pages are a valuable resource,
that online html documentation is also a good thing to have, and
that info pages are way \fIway\fP ahead of their time\&. The
\fBNOTES\fP section explains how this page was created\&.
.SH OPTIONS
.nr mi \n(.iu
.TP
\fB-I\fP f (\fIinflation\fP)
Sets the main inflation value to f\&. This value is the main handle
for affecting cluster granularity\&. It is usually chosen somewhere
in the range [1\&.2-5\&.0]\&. \fB-I\fP\ \fB5\&.0\fP will tend to result
in fine-grained clusterings, and \fB-I\fP\ \fB1\&.2\fP will tend to
result in very coarse grained clusterings\&. Your mileage will vary
depending on the characteristics of your data\&. That is why it is
a good idea to test the quality and coherency of your clusterings
using \fBclmdist(1)\fP and \fBclminfo(1)\fP\&. This will most likely reveal that
certain values of \fB-I\fP are simply not right for your data\&. The
\fBclmdist(1)\fP section contains a discussion of how to use the cluster
validation tools shipped with \fBmcl\fP (see the \fBSEE ALSO\fP section)\&.

A second option for affecting cluster granularity is the
\fB-c\fP option\&.
It may possibly increase granularity\&.

With low values for \fB-I\fP, like \fB-I\fP\ \fB1\&.2\fP, you should be
prepared to use more resources in order to maintain quality of
clusterings, i\&.e\&. increase the arguments to the
\fB-P/-S/-R\fP options or simply increase
the argument to \fB-scheme\fP\&.
.TP
\fB-o\fP str (\fIfname\fP)
Output the clustering to file named fname\&. The default file name
is out\&.mcl\&. It is possible to send the clustering to stdout
by supplying \fB-o\fP\ \fB-\fP\&. The clustering is output in the
mcl matrix format; see the \fBmcx(5)\fP section for
more information on this\&.
.TP
\fB-c\fP f (\fIcentering\fP)
The larger the value of f the more nodes are attached
to themselves rather than their neighbours, the more
expansion (the spreading of flow through the graph) is
opposed, and the more fine-grained clusterings tend to be\&. f should be
chosen greater than or equal to 1\&.0\&. The default is f=1\&.0\&. This option
has a much weaker effect than the \fB-I\fP option, but it can be
useful depending on your data\&.

The following resource control options are discussed as a group:
.TP
\fB-p\fP f (\fIcutoff\fP)
.TP
\fB-P\fP n (\fI1/cutoff\fP)
.TP
\fB-S\fP s (\fIselection number\fP)
.TP
\fB-R\fP r (\fIrecover number\fP)
.TP
\fB-pct\fP pct (\fIrecover percentage\fP)
.TP
\fB-scheme\fP k (\fIuse a preset pruning scheme\fP)
.TP
\fB--show-schemes\fP (\fIshow preset schemes\fP)
If you do not know nor want to know the mcl internals, just ignore the
additional information below and follow the recommendations\&. They are
quite straightforward\&.

After computing a new (column stochastic) matrix vector during expansion
(which is matrix multiplication c\&.q\&. squaring), the vector is
successively exposed to different pruning strategies\&. The intent of
pruning is that many small entries are removed while retaining much of
the stochastic mass of the original vector\&. After pruning, vectors are
rescaled to be stochastic again\&. MCL iterands are theoretically known to
be sparse in a weighted sense, and this manoever effectively perturbs the
MCL process a little in order to obtain matrices that are genuinely
sparse, thus keeping the computation tractable\&. An example of monitoring
pruning can be found in the discussion of \fB-v\fP\ \fBpruning\fP under the
\fB--verbose\fP option\&.

\fBmcl\fP proceeds as follows\&.
First, entries that are smaller than \fIcutoff\fP are removed,
resulting in a vector with at most \fI1/cutoff\fP entries\&.
The cutoff can be supplied either by \fB-p\fP, or as the inverse value
by \fB-P\fP\&. The latter is more intuitive, if your intuition is like
mine (and the P stands for precision or pruning by the way)\&.
The cutoff just described is rigid; it is the same for all vectors\&. The
\fB--adapt\fP option causes the computation of a
cutoff that depends on a vector\&'s homogeneity properties, and this option
may speed up \fBmcl\fP considerably\&.

Second, if the remaining stochastic mass (i\&.e\&. the sum of all remaining
entries) is less than \fIpct\fP/100 and the number of remaining
entries is less than \fIr\fP (as specified by the \fB-R\fP flag),
\fBmcl\fP will try to regain ground by recovering the largest discarded
entries\&. The total number of entries is not allowed to grow larger than
\fIr\fP\&.
If recovery was not necessary, \fBmcl\fP tries to prune the vector further
down to at most \fIs\fP entries (if applicable), as specified by the
\fB-S\fP flag\&. If this results in a vector that satisfies the recovery
condition then recovery is attempted, exactly as described above\&.
The latter will not occur of course if \fIr\fP <= \fIs\fP\&.

The default setting is something like \fB-P\fP\ \fB2000\fP \fB-S\fP\ \fB500\fP
\fB-R\fP\ \fB600\fP\&. Check the \fB-z\fP flag to be sure\&. There is a set
of precomposed settings, which can be triggered with the
\fB-scheme\fP\ \fIk\fP option\&. \fIk\fP=2 is the default
scheme; higher values for \fIk\fP result in costlier and
more accurate computations (vice versa for lower, cheaper, and less
accurate)\&. The schemes are listed using the
\fB--show-schemes\fP option\&. It is advisable to use the
\fB-scheme\fP option only in interactive mode, and to use the explicit
expressions when doing batch processing\&. The reason is that there
is \fIno guarantee whatsoever\fP that the schemes will not change
between different releases\&. This is because the scheme options
should reflect good general purpose settings, and it may become
appararent that other schemes are better\&.

Note that \&'less accurate\&' or \&'more accurate\&' computations
may still generate the same output clusterings\&. Use \fBclmdist\fP
to compare output clusterings for different resource parameters\&.
Refer to \fBclmdist(1)\fP for a discussion of this issue\&.

As a reminder of the existence of pruning and its importance for both
speed and accuracy, \fBmcl\fP reports three numbers when it is done, the
\&'jury marks\&'\&.
These are somewhat (but not totally) indicative for the quality of
pruning, and they are excerpts from the output produced by
\fB-v\fP\ \fBpruning\fP, namely the numbers listed under the \fInx\fP column
for the three first iterations\&. Each number gives the average stochastic
mass of the \fBnx\fP worst instances of pruning, i\&.e\&. those vectors
for which the most mass was removed\&. The average is listed as a
percentage\&. The numbers should preferably be higher than 70\&. If they are
in the vicinity of 80 or 90, \fBmcl\fP is doing fine as far as pruning is
concerned\&. Choose a higher scheme if you think them too low\&.
For very dense graphs that do have strong cluster structure,
the jury marks can sink as low as to the 30\&'s
and 40\&'s, but the clusterings generated by \fBmcl\fP may still be good\&.
Refer to the \fB-v\fP option for more information, and
note that the jury becomes friendlier, resp\&. harsher when the
\fB-nx\fP option is increased/decreased\&.
.TP
\fB-warn-pct\fP k (\fIprune warn percentage\fP)
.TP
\fB-warn-factor\fP k (\fIprune warn factor\fP)
The two options \fB-warn-pct\fP and \fB-warn-factor\fP relate to
warnings that may be triggered once the \fIinitial\fP pruning of a vector
is completed\&. The idea is to issue warnings if initial pruning almost
completely destroys a computed vector, as this may be a sign that the
pruning parameters should be changed\&. It depends on the mass remaining
after initial pruning whether a warning will be issued\&. If that mass is
less than \fIwarn-pct\fP or if the number of remaining entries is smaller
by a factor \fIwarn-factor\fP than both the number of entries originally
computed \fIand\fP the recovery number, in that case, \fBmcl\fP will issue a
warning\&.

\fB-warn-pct\fP takes an integer between 0 and 100 as parameter,
\fB-warn-factor\fP takes a real positive number\&. They default to
something like 30 and 50\&.0\&. If you want to see less warnings, decrease
\fIwarn-pct\fP and increase \fIwarn-factor\fP\&. Set \fIwarn-factor\fP to zero
if you want no warnings\&.
.TP
\fB--rigid\fP (\fIpruning\fP)
See the \fB--adapt\fP option below\&.
.TP
\fB-ae\fP f (\fIadaptive pruning exponent\fP)
See the \fB--adapt\fP option below\&.
.TP
\fB-af\fP f (\fIadaptive pruning factor\fP)
See the \fB--adapt\fP option below\&.
.TP
\fB--adapt\fP (\fIpruning\fP)
The default \fBmcl\fP pruning behaviour as described under

family/src/mcl/mcl/mcl.c
/*
// mcl.c                Core MCL clustering algorithm
*/

#include <pthread.h>

#include "mcl/mcl.h"
#include "mcl/params.h"
#include "mcl/dpsd.h"

#include "nonema/io.h"

#include "util/txt.h"
#include "util/file.h"
#include "util/types.h"
#include "util/alloc.h"


#define ITERATION_INITIAL  1
#define ITERATION_MAIN     2


static int mclVerbosityStart    =  1;

static void pruningExplain
(  FILE* vbfp
)  ;

typedef struct
{
   int            id
;  int            start
;  int            end
;  float          power
;  mcxMatrix*     mx
;
}  mcxFlowInflate_thread_arg   ;


pthread_t         *threads_inflate;
pthread_t         *threads_expand;              /* used in mcl/compose.c */
static pthread_attr_t  pthread_custom_attr;


void mcxFlowInflate_threaded
(  
   mcxMatrix*        mx
,  float             power
)  ;


void  mcxFlowInflate_thread
(  
   void *arg
)  ;


void  mcxDumpMatrix
(  
   mcxMatrix*     mx
,  const char*    fstem
,  const char*    affix
,  const char*    pfix
,  int            n
,  int            printValue
)  ;


void  mcxDumpVector
(  
   mcxVector*     vec
,  const char*    fstem
,  const char*    affix
,  const char*    pfix
,  int            n
,  int            printValue
)  ;


int mcxDoIteration
(  
   mcxMatrix**       mxin
,  mcxMatrix**       mxout
,  const mclParam*   param
,  mcxComposeStats*  stats
,  int               type
,  int               ite_index
)  ;


mclParam* mclParamNew
(  void
)  
   {  
      mclParam* param = (mclParam*) rqAlloc(sizeof(mclParam), EXIT_ON_FAIL)

   ;  param->mainInflation             =     2
   ;  param->mainLoopLength            =     10000
   ;  param->initInflation             =     2
   ;  param->initLoopLength            =     0
                                     
   ;  param->printDigits               =     3
   ;  param->printMatrix               =     0
                                     
   ;  param->mcxIpretParam             =     mcxIpretParamNew()
   ;  return param
;  }


mcxMatrix*  mclCluster
(  
   mcxMatrix*        mxEven
,  const mclParam*   param
)
   {
      mcxComposeStats*  stats          =  mcxComposeStatsNew(mcl_nx,mcl_ny)

   ;  mcxMatrix*        mxOdd          =  NULL
   ;  mcxMatrix*        mxCluster      =  NULL
   ;  int               i              =  0
   ;  int               n_vec          =  mxEven->N_cols
   ;  int               bPrint         =  param->printMatrix
   ;  int               digits         =  param->printDigits

   ;  threads_inflate
      =  (pthread_t *) rqAlloc
         (mcl_num_ithreads*sizeof(pthread_t), EXIT_ON_FAIL)

   ;  threads_expand
      =  (pthread_t *) rqAlloc
         (mcl_num_ethreads*sizeof(pthread_t), EXIT_ON_FAIL)

   ;  if (bPrint)
      mcxFlowPrettyPrint
      (  mxEven   ,  stdout
      ,  digits   ,  "1 After centering (if) and normalization"
      )

   ;  if (mclDumpIterands)
      mcxDumpMatrix(mxEven, mclDumpStem, "ite", "", 0, 1)

   ;  if (mclVerbosityVectorProgress)
      {  
         if (!mclVerbosityPruning)
         fprintf(stdout, " ite ")

      ;  for (i=0;i<n_vec/mclVectorProgression;i++)
         fputc('-', stdout)

      ;  fprintf(stdout, mclVerbosityPruning ? "\n" : " inhomogeneity\n")
   ;  }
      else if (mclVerbosityMatrixProgress)
      fprintf(stdout, " ite  inhomogeneity\n")

   ;  for (i=0;i<param->initLoopLength;i++)
      {  
         int convergence = mcxDoIteration (  &mxEven
                                          ,  &mxOdd
                                          ,  param
                                          ,  stats
                                          ,  ITERATION_INITIAL
                                          ,  i+1
                                          )
      ;  convergence    *= 1              /* don't care for initial loop */
      ;  mcxMatrixFree(&mxEven)
      ;  mxEven  =  mxOdd
   ;  }

      if (  param->initLoopLength
         && (  mclVerbosityPruning
            || mclVerbosityVectorProgress
            )
         )
      fprintf
      (  stdout
      ,  "====== Changing from initial to main inflation now ====\n"
      )

   ;  for (i=0;i<param->mainLoopLength;i++)
      {  
         int convergence = mcxDoIteration (  &mxEven
                                          ,  &mxOdd
                                          ,  param
                                          ,  stats
                                          ,  ITERATION_MAIN
                                          ,  i+1+param->initLoopLength
                                          )
      ;  mcxMatrixFree(&mxEven)
      ;  mxEven  =  mxOdd
      ;  if (convergence)
         {  if (mclVerbosityPruning)
            fprintf(stdout, "\n")
         ;  break
      ;  }
   ;  }

   ;  mxCluster = mcxInterpret(mxEven, param->mcxIpretParam)
   ;  mcxMatrixFree(&mxEven)

   ;  rqFree(threads_inflate)
   ;  rqFree(threads_expand)

   ;  mcxComposeStatsFree(stats)

   ;  return mxCluster
;  }

void mcxMatrixCenter
(  mcxMatrix*        mx
,  float             w_center
,  float             w_selfval
)  {  
      int      col

   ;  for (col=0;col<mx->N_cols;col++)
      {
         mcxVector*  vec      =  mx->vectors+col
      ;  mcxIvp*     match    =  NULL
      ;  int         offset   =  -1
      ;  float       selfval

      ;  if
         (  vec->ivps
         && (  mcxVectorIdxVal(vec, col, &offset)
            ,  offset >= 0
            )
         )  
         {  
            match = (vec->ivps+offset)
         ;  selfval           =  match->val
         ;  match->val        =  0.0
      ;  }
         else                    /* create extra room in vector */
         { 
            mcxVectorResize (vec, (vec->n_ivps)+1)
         ;  match             =  vec->ivps+(vec->n_ivps-1)
         ;  match->val        =  0.0
         ;  match->idx        =  col
         ;  mcxVectorSort (vec, mcxIvpIdxCmp)
                                 /* ^^^ this could be done by shifting */

         ;  mcxVectorIdxVal(vec, col, &offset)

         ;  if (offset < 0)
            fprintf
            (  stderr
            ,  "[mcxMatrixCenter SNH] insertion failed ??\n"
            )
         ,  exit(1)

         ;  match             =  (vec->ivps+offset)
         ;  selfval           =  0.0
      ;  }

         {  
            float sum            =  mcxVectorSum(vec)

         ;  if (sum == 0.0)
            {  match->val =  1
         ;  }
            else
            {  match->val        =  w_center == 0.0
                                    ?  0.0
                                    :  (  w_center * mcxVectorPowSum(vec, 2)
                                          +  w_selfval * selfval
                                       ) / sum
         ;  }
      ;  }

      ;  mcxVectorNormalize(vec)
   ;  }
;  }


void mcxFlowInflate
(  
   mcxMatrix*              mx
,  float                   power
)  {  
      mcxVector*     vecPtr          =     mx->vectors
   ;  mcxVector*     vecPtrMax       =     vecPtr + mx->N_cols

   ;  while (vecPtr < vecPtrMax)
      {  mcxVectorInflate(vecPtr, power)
      ;  vecPtr++
   ;  }
;  }


void mcxFlowInflate_threaded
(  
   mcxMatrix*        mx
,  float             power
)  {
      mcxVector*     vecPtr      =  mx->vectors
   ;  mcxVector*     vecPtrMax   =  vecPtr + mx->N_cols
   ;  int            workLoad    =  mx->N_cols / mcl_num_ithreads
   ;  int            workTail    =  mx->N_cols % mcl_num_ithreads
   ;  int            i           =  0

   ;  pthread_attr_init(&pthread_custom_attr)

   ;  for (i=0;i<mcl_num_ithreads;i++)
      {
         mcxFlowInflate_thread_arg *a
                     =  (mcxFlowInflate_thread_arg *)
                        malloc(sizeof(mcxFlowInflate_thread_arg))
      ;  a->id       =  i
      ;  a->start    =  workLoad * i
      ;  a->end      =  workLoad * (i+1)
      ;  a->mx       =  mx
      ;  a->power    =  power

      ;  if (i+1==mcl_num_ithreads)
         a->end   +=  workTail

      ;  pthread_create
         (  &threads_inflate[i]
         ,  &pthread_custom_attr
         ,  (void *) mcxFlowInflate_thread
         ,  (void *) a
         )
   ;  }

   ;  for (i = 0; i < mcl_num_ithreads; i++)
      pthread_join(threads_inflate[i], NULL)
;  }


void  mcxFlowInflate_thread
( 
   void *arg
)
   {
      mcxFlowInflate_thread_arg
                     *a          =  (mcxFlowInflate_thread_arg *)  arg
   ;  mcxMatrix*     mx          =  a->mx
                     
   ;  mcxVector*     vecPtr      =  mx->vectors + a->start
   ;  mcxVector*     vecPtrMax   =  mx->vectors + a->end
   ;  float          power       =  a->power

   ;  while (vecPtr < vecPtrMax)
      {
         mcxVectorInflate(vecPtr, power)
      ;  vecPtr++
   ;  }

   ;  free(a)
;  }


int mcxDoIteration
(  mcxMatrix**          mxin
,  mcxMatrix**          mxout
,  const mclParam*      param
,  mcxComposeStats*     stats
,  int                  type
,  int                  ite_index
)  {
      int               bPrint         =  param->printMatrix
   ;  int               digits         =  param->printDigits
   ;  FILE*             vbfp           =  stdout
   ;  int               bInitial       =  (type == ITERATION_INITIAL)
   ;  const char        *when          =  bInitial ? "initial" : "main"
   ;  float             inflation      =  bInitial
                                          ?  param->initInflation
                                          :  param->mainInflation
   ;  char              msg[80]

   ;  if (mclVerbosityVectorProgress && !mclVerbosityPruning)
      fprintf(stdout, "%3d  ", ite_index)

   ;  if (  (mclModeCompose == MCL_COMPOSE_DENSE)
         && (ite_index > 1)
         && mcxMatrixNrofEntries(*mxin) < pow((*mxin)->N_cols, 1.5)
         )
      mclModeCompose = MCL_COMPOSE_SPARSE

   ;  *mxout =  mcxFlowExpand(*mxin, stats )

   ;  if (bPrint)
      {  sprintf
         (  msg, "%d%s%s%s"
         ,  2*ite_index, " After mcxFlowExpand (", when, ")"
         )
      ;  mcxFlowPrettyPrint(*mxout, stdout, digits, msg)
   ;  }


   ;  if (mclVerbosityVectorProgress)
      {  if (mclVerbosityPruning)
         fprintf(stdout, "\n")
      ;  else
         fprintf(stdout, " %6.2f\n", stats->inhomogeneity)
   ;  }
      else if (!mclVerbosityPruning && mclVerbosityMatrixProgress)
      fprintf(stdout, "%3d  %6.2f\n", ite_index, stats->inhomogeneity)


   ;  if (  mclVerbosityPruning || mclDumpClusters )
      { 
         mcxMatrix*  clus  =  mcxInterpret(*mxout, param->mcxIpretParam)

      ;  if (mclVerbosityPruning)
         {  
            if (mclVerbosityStart)
            {  
               pruningExplain(vbfp)
            ;  mclVerbosityStart = 0
         ;  }

         ;  mcxComposeStatsPrint(stats, vbfp)
         ;  fprintf(vbfp, "\n")
      ;  }

      ;  if (mclDumpClusters)
         {  mcxDumpMatrix(clus, mclDumpStem, "cls", "", ite_index, 0)
      ;  }
      ;  mcxMatrixFree(&clus)
   ;  }

      if (mcl_num_ithreads)
      mcxFlowInflate_threaded(*mxout, inflation)
   ;  else
      mcxFlowInflate(*mxout, inflation)

   ;  if (bPrint)
      {  sprintf
         (  msg,  "%d%s%s%s"
         ,  2*ite_index+1, " After inflation (", when, ")"
         )
      ;  mcxFlowPrettyPrint(*mxout, stdout, digits, msg)
   ;  }

   ;  if (mclDumpIterands)
         mcxDumpMatrix(*mxout, mclDumpStem, "ite", "", ite_index, 1)

   ;  if (mclDumpAttractors)
      {  mcxMatrix*  diago = mcxDiagOrdering(*mxout, &mcl_vec_attr)
      ;  mcxDumpVector(mcl_vec_attr, mclDumpStem, "attr", ".vec", ite_index, 1)
      ;  mcxMatrixFree(&diago)
   ;  }

   ;  if (stats->inhomogeneity < mclInhomogeneityStop)
      return 1
   ;  else
      return 0
;  }


void  mcxDumpMatrix
(  mcxMatrix*     mx
,  const char*    fstem
,  const char*    affix
,  const char*    postfix
,  int            n
,  int            printValue
)  {  mcxIOstream*   xfDump
   ;  char snum[18]
   ;  mcxTxt*  fname

   ;  if (  (  mclDumpOfset
            && (n<mclDumpOfset)
            )
         || (  mclDumpBound
            && (n >= mclDumpBound)
            )
         )
         return

   ;  fname = mcxTxtNew(fstem)
   ;  mcxTxtAppend(fname, affix)

   ;  sprintf(snum, "%d", n)
   ;  mcxTxtAppend(fname, snum)
   ;  mcxTxtAppend(fname, postfix)

   ;  xfDump   =  mcxIOstreamNew(fname->str, "w")

   ;  if (mcxIOstreamOpen(xfDump, RETURN_ON_FAIL) != STATUS_OK)
      {  fprintf
         (  stderr
         ,  "[mcxDumpMatrix] cannot open stream [%s], ignoring\n"
         ,  xfDump->fn->str
         )
      ;  return
   ;  }
      else
      {  if (mclDumpMode == 'a')
         mcxMatrixWriteAscii(mx, xfDump, printValue ? 8 : -1, RETURN_ON_FAIL)
      ;  else
         mcxMatrixWrite(mx, xfDump, RETURN_ON_FAIL)
   ;  }

   ;  mcxIOstreamFree(&xfDump)
   ;  mcxTxtFree(&fname)
;  }


void  mcxDumpVector
(  mcxVector*     vec
,  const char*    fstem
,  const char*    affix
,  const char*    postfix
,  int            n
,  int            printValue
)  {  mcxIOstream*   xf
   ;  char snum[18]
   ;  mcxTxt*  fname

   ;  if (  (  mclDumpOfset
            && (n<mclDumpOfset)
            )
         || (  mclDumpBound
            && (n >= mclDumpBound)
            )
         )
         return

   ;  fname = mcxTxtNew(fstem)
   ;  mcxTxtAppend(fname, affix)

   ;  sprintf(snum, "%d", n)
   ;  mcxTxtAppend(fname, snum)
   ;  mcxTxtAppend(fname, postfix)

   ;  xf =  mcxIOstreamNew(fname->str, "w")
   ;  if (mcxIOstreamOpen(xf, RETURN_ON_FAIL) == STATUS_FAIL)
      {  mcxTxtFree(&fname)
      ;  mcxIOstreamFree(&xf)
      ;  return
   ;  }

   ;  if (mclDumpMode == 'a')
      mcxVectorWriteAscii(vec, xf->fp, printValue ? 8 : -1)
   ;  else
      mcxVectorWrite(vec, xf, RETURN_ON_FAIL)

   ;  mcxIOstreamFree(&xf)
   ;  mcxTxtFree(&fname)
;  }


static void pruningExplain
(  FILE* vbfp
)  
   {
   const char* pruneMode      =  (mclModePruning == MCL_PRUNING_ADAPT)
                              ?  "adaptive"
                              :  "rigid"  
;  const char* pStatus        =  (mclModePruning == MCL_PRUNING_ADAPT)
                              ?  "minimum"
                              :  "rigid"  
;

fprintf
(  vbfp
,  "The %s threshold value equals [%f] (-p <f> or -P <i>)).\n"
,  pStatus
,  mclPrecision
)  ;

if (mclMarknum && mclSelect)
fprintf
(  vbfp
,  "Exact selection prunes to at most %d positive entries (-m <i>).\n"
,  mclMarknum
)  ;

if (mclMarknum && mclRecover)
fprintf
(  vbfp
,  "If %s thresholding leaves less than %2.0f%% mass (-pct) and less than\n"
   "%d entries, as much mass as possible is recovered (--recover).\n"
,  pruneMode
,  100 * mclPct
,  mclMarknum
)  ;

fprintf
(  vbfp
,  "\nLegend\n"
   "all: average over all vectors\n"
   "ny:  average over the worst %d cases (-ny <i>).\n"
   "nx:  average over the worst %d cases (-nx <i>).\n"
   "<>:  recovery is %sactivated. (-m, -M, -pct, --recover)\n"
   "[]:  selection is %sactivated. (-m, -M, --select)\n"
   "recovery and selection: 'mcldoc pruning' explains\n"
   "8532c:  'mcldoc 8532c' explains.\n"
,  mcl_ny
,  mcl_nx
,  (mclMarknum && mclRecover) ? "" : "NOT "
,  (mclMarknum && mclSelect) ? "" : "NOT "
)  ;

fprintf
(  vbfp
,  "\n"
"----------------------------------------------------------------------------\n"
" mass percentages  | distributions of vector sizes |#recover cases <>       \n"
"         |         |__ compose ________ prune _____||    #select cases []   \n"
"  prune  | final   |000  00   0    |000  00   0    ||       |     #below pct\n"
"all ny nx|all ny nx|8532c8532c8532c|8532c8532c8532c|V       V        V      \n"
"---------.---------.---------------.---------------.-------.--------.-------\n"
)  ;
}


